### ISSUE TITLE
Intersection Observer review

### ISSUE THREAD

**Issue by LJWatson** [w3ctag/design-reviews#197]
The WebPlat WG has just published the FPWD of [Intersection Observer](https://www.w3.org/TR/intersection-observer/), triggering wide review. Thanks to its time in incubation, the spec already has good interop, so we're hopeful of moving to CR reasonably quickly.

We'd welcome a review by the TAG. Feedback no later than Friday 27th October would be appreciated, feedback earlier even more so.

Please file issues on the [Intersection Observer repo](https://github.com/w3c/intersectionobserver/issues/new). Thank you.

**Comment by dbaron** [comment]
So one thing I need to investigate further is that it feels to me that intersection observer running before the animation frame callbacks seems wrong, I think it should run substantially after them, with a bunch of other stuff in between, since rAF is primarily a writing callback (so should be early, before the natural style/layout), whereas IntersectionObserver is primarily a reading callback (so should be late, after the natural style/layout).  But I suspect that some of my concern there may be that I think where HTML describes the running of the animation frame callbacks being wrong.  I think this requires a good bit of digging into (to see what current specs say, and how interoperable existing implementations are); it's related to our #72.

**Comment by plinss** [comment]
I have concerns about the rootMargin being a string based API. This should be using [CSS Typed OM](https://drafts.css-houdini.org/css-typed-om/) if not now, at least some point if the future. The spec should at a minimum have a note about a typed OM version of the margin.

**Comment by LJWatson** [comment]
Gentle ping to remind you that we're looking for issues to be filed by the end of next week. Thanks.

**Comment by torgo** [comment]
Discussed on [today's telecon](https://cryptpad.fr/pad/#/1/view/GpGiv+gMMGEMUdb6lQ2akg/gCtzEJ3B8wHxZ2lPsbTKg2ffmacIq3g5Ljd3ZSi80-0/).

**Comment by dbaron** [comment]
I'd like to leave this open to monitor the progress on w3c/IntersectionObserver#263 and whatwg/html#3246 etc.

**Comment by slightlyoff** [comment]
Per today's F2F, now for me to discuss open issues with @szager-chromium

**Comment by slightlyoff** [comment]
Pinged @szager-chromium; unclear where things are now. Will circle back.

**Comment by szager-chromium** [comment]
I want to hopefully close the loop on the comment by @dbaron:

> So one thing I need to investigate further is that it feels to me that intersection observer running before the animation frame callbacks seems wrong, I think it should run substantially after them, with a bunch of other stuff in between, since rAF is primarily a writing callback (so should be early, before the natural style/layout), whereas IntersectionObserver is primarily a reading callback (so should be late, after the natural style/layout).

You are quite right that IntersectionObserver should run after rAF callbacks; I just created a [pull request](https://github.com/w3c/IntersectionObserver/pull/331l) for that change. The original language there was a mistake.

This sidesteps the issue of when exactly style/layout/paint gets updated, but that issue is better addressed in the separate spec issues you called out.

**Comment by dbaron** [comment]
Yep, that PR improves things, thanks.

**Comment by dbaron** [comment]
... but I think, as I mentioned in https://github.com/w3c/IntersectionObserver/issues/263#issuecomment-448418727, that the steps should be still later.  I think this requires writing a bunch of tests to check what happens today, though...

**Comment by LJWatson** [comment]
@szager @dbaron what is the state of this now?

**Comment by dbaron** [comment]
We separated out the more general bits into #489, but we're perhaps still leaving this open to wait on w3c/IntersectionObserver#263.

**Comment by dbaron** [comment]
@plinss and I are looking at this in a breakout in our virtual "Cork" face-to-face.

So we were leaving this open for w3c/IntersectionObserver#263, which is in turn waiting on whatwg/html#3246, which is in turn waiting on w3c/csswg-drafts#5115.  While it would be nice to fix that sequence of things, we probably don't need to leave this review open for it.

These are also all referenced from #489.

**Comment by dbaron** [comment]
OK, we discussed in mini-plenary and we're happy closing this issue at this point.  Thanks for requesting TAG review... a long time ago.  Sorry this took so long.

**Comment by LJWatson** [comment]
Thank you @dbaron  and @plinss.
